### PR TITLE
XACKDEL Command

### PR DESCRIPTION
**Pull request by nickiaq**
This PR adds a stream command `XACKDEL` that acknowledges one or more messages and (conditionally) deletes them from the stream. The command is especially valuable for teams using Valkey streams as a Kafka alternative. This implementation is compatible with the equivalent command introduced in Redis 8.2.0.

For more background on why `XACKDEL` is necessary, see #2903. [This article](https://blog.arcjet.com/replacing-kafka-with-redis-streams/) describes a common pattern when processing a stream in Valkey's current state. To cleanup stream message, a separate "janitor" process tracks each consumer group's progress and periodically calls `XTRIM` to reclaim space once all messages up to some high water mark have been ack'd by all groups.

Another alternative is to use a Lua script that replicates the `XACKDEL` logic. Consumer groups can use the script to acknowledge and (conditionally) delete each batch of messages. But, iterating over consumer groups in Lua is slower. And in both the Lua script solution and the janitor process method, there's the risk that the developer introduces a race condition or logic error leading to an unbounded accumulation of stream messages.

`XACKDEL` with `ACKED` mode makes this janitor process or Lua script unnecessary.

For full Redis compatibility, the command supports the same three deletion modes with the same semantics, command syntax, and reply format:

- `KEEPREF` (default, implicit) acks and deletes immediately, leaving PEL references in other groups
- `ACKED` only deletes once every consumer group has acknowledged or passed the message, making it safe for fan-out stream topologies
    - Here, if the message was ack'd but not deleted, the response is 2
- `DELREF` acks, deletes, and forcibly removes PEL entries from all other groups

I've got tests for each mode and possible state included along with other tests for replication, syntax, and other edge cases.

This is my first PR (alongside `XDELEX`). And I've tried to follow the style guide & contributing guidelines. Happy to make any corrections if I've missed something.

It's possible to create a shared function for parsing the mode & id count between `XACKDEL` and `XDELEX`. So I can follow up with a PR to reduce that duplicated code. Or, I can merge these 2 PRs into one.

### CODE REVIEW

**Comment by nickiaq** [comment]
@zuiderkwast Thanks for updating this to the latest `unstable`. If it helps for me to rebase or make any other changes, I'm happy to. (Same for #3467)

**Comment by zuiderkwast** [comment]
> @zuiderkwast Thanks for updating this to the latest `unstable`. If it helps for me to rebase or make any other changes, I'm happy to. (Same for #3467)

I did it just to trigger the "Provenance Guard" CI job. It checks for code from Redis, which is under an incompatible license. It looks green. 👍

I hope we'll get some time to review it in the next weeks. Now, we're finishing the 9.1 release, so this will target the next version.

**Comment by nickiaq** [comment]
Thanks for the update. I'll watch out for any review comments to address whenever you have a chance to look.

I don't see a 9.2 project on the projects tab, just 10. So it looks like `"since": "10.0.0"` is right for that in [src/commands/xackdel.json](https://github.com/valkey-io/valkey/pull/3466/changes#diff-f017582668b80ef3dadf9a75ffe4ac53e71645540808773121d4a91c1b05c72a). But, that's an easy change if there's a 9.1.1 or 9.2 coming.

**Comment by nickiaq** [comment]
The force push a few seconds ago was just keeping this up with unstable.

File: src/commands/xackdel.json [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+                        "type": "pure-token",
+                        "token": "DELREF"
+                    },
+                    {
+                        "name": "acked",
+                        "type": "pure-token",
+                        "token": "ACKED"
+                    }
+                ]
+            },
```
[the remark this answers is not in the document]
Comment (nickiaq) [reply src/commands/xackdel.json:69]: Fixed in [e3ed4d8](https://github.com/valkey-io/valkey/pull/3466/commits/e3ed4d8336aeae18b7bb1aaf72fc14567d32d16b)

File: tests/unit/type/stream.tcl [hunk 2]
```diff
@@ -634,6 +634,405 @@ start_server {
...
+
+        # XACKDEL clears the PEL and still returns 1 even though already deleted
+        set ids [r XACKDEL testxadstream testxadgrp1 ACKED IDS 1 1-0]
+        assert_equal 1 [llength $ids]
+        assert_equal 1 [lindex $ids 0]
+
+        # PEL must be cleared despite the -1 response code
+        set pend [r XPENDING testxadstream testxadgrp1]
+        assert_equal 0 [lindex $pend 0]
+    }
```
[the remark this answers is not in the document]
Comment (nickiaq) [reply tests/unit/type/stream.tcl:1063]: Fixed in [e3ed4d8](https://github.com/valkey-io/valkey/pull/3466/commits/e3ed4d8336aeae18b7bb1aaf72fc14567d32d16b)

File: src/t_stream.c [hunk 3]
```diff
@@ -3558,6 +3558,262 @@ void xdelCommand(client *c) {
...
+            streamGetEdgeID(s, 1, 1, &s->first_id);
+        }
+    }
+
+    /* Propagate the write if needed. */
+    if (deleted) {
+        signalModifiedKey(c, c->db, c->argv[1]);
+        notifyKeyspaceEvent(NOTIFY_STREAM, "xdel", c->argv[1], c->db->id);
+        server.dirty += deleted;
+    }
```
[the remark this answers is not in the document]
Comment (nickiaq) [reply src/t_stream.c:3821]: I believe this review comment is not correct from the AI. It made good catches on the others, especially [this one from the PR for XDELEX](https://github.com/valkey-io/valkey/pull/3467#discussion_r3471105368). But, I think using keyspace notification `xdel` is valid here, because it is similar to other usage like `del`, which is emitted in many commands not just `DEL`. For example, `GETDEL` emits `del`. So, I'm thinking that `xdel`  is the comparable event to emit here to match `XDEL`.

Happy to correct if someone knows better though.

**Comment by hpatro** [comment]
@valkey-review-bot Please review this change.

File: src/t_stream.c [hunk 4]
```diff
@@ -3558,6 +3558,262 @@ void xdelCommand(client *c) {
...
+    } else if (strcasecmp(objectGetVal(c->argv[argi]), "DELREF") == 0) {
+        argi += 1;
+        mode = 1;
+    } else if (strcasecmp(objectGetVal(c->argv[argi]), "ACKED") == 0) {
+        argi += 1;
+        mode = 2;
+    } else if (strcasecmp(objectGetVal(c->argv[argi]), "IDS") != 0) {
+        addReplyError(c, "xackdel mode not recognized");
+        return;
+    }
```
Comment (hpatro) [inline src/t_stream.c:3593]: It doesn't have to be `IDS` to be invalid.
```suggestion
    } else {
        addReplyError(c, "xackdel mode not recognized");
        return;
    }
```

File: src/t_stream.c [hunk 5]
```diff
@@ -3558,6 +3558,262 @@ void xdelCommand(client *c) {
...
+    argi++; /* past IDS */
+
+    /* Parse and validate numids: must be a positive integer. */
+    long long id_count;
+    if (getLongLongFromObject(c->argv[argi], &id_count) != C_OK || id_count <= 0) {
+        addReplyError(c, "Number of IDs must be a positive integer");
+        return;
+    }
```
Comment (hpatro) [inline src/t_stream.c]: Would prefer splitting the parsing in two parts like we do for other flows here to get consistent error message across stream commands.
```suggestion
    if (getLongLongFromObjectOrReply(c->argv[argi], &id_count) == C_ERR) {
        return;
    }
    if (id_count <= 0) {
        addReplyError(c, "The IDs argument must be a positive integer");
        return;
    }
```

File: src/t_stream.c [hunk 6]
```diff
@@ -3558,6 +3558,262 @@ void xdelCommand(client *c) {
...
+    if (id_count > actual_ids) {
+        addReplyError(c, "numids parameter must match the number of IDs provided");
+        return;
+    } else if (id_count < actual_ids) {
+        addReplyError(c, "syntax error");
+        return;
+    }
+
+    /* If missing stream or group, return -1 for each ID. */
+    if (o == NULL || group == NULL) {
```
Comment (hpatro) [inline src/t_stream.c:3624]: `o` can't be null given we have already checked above. 
```suggestion
    if (group == NULL) {
```

File: src/t_stream.c [hunk 7]
```diff
@@ -3558,6 +3558,262 @@ void xdelCommand(client *c) {
...
+    }
+
+    int acked = 0;
+    int deleted = 0;
+    int first_entry = 0;
+
+    /* Fast path for KEEPREF. Since we only need to cleanup the PEL for the target
+     * group, we only need to loop over messages (and not consumers) and can set
+     * responses inline. Thus, there's a separate setup for KEEPREF vs. ACKED/DELREF*/
+    if (mode == 0 /* KEEPREF */) {
```
Comment (hpatro) [inline src/t_stream.c]: Nit:
```suggestion
    if (mode == 0) { /* KEEPREF */
```

File: src/t_stream.c [hunk 8]
```diff
@@ -3558,6 +3558,262 @@ void xdelCommand(client *c) {
...
+                void *result;
+                if (raxFind(cg->pel, buf, sizeof(buf), &result)) {
+                    if (is_target || mode == 1 /* DELREF */) {
+                        /* Remove from this group's PEL (target always; DELREF removes all). */
+                        streamNACK *nack = result;
+                        raxRemove(cg->pel, buf, sizeof(buf), NULL);
+                        raxRemove(nack->consumer->pel, buf, sizeof(buf), NULL);
+                        streamFreeNACK(nack);
+                        acked++;
+                    } else if (mode == 2 /* ACKED */) {
```
Comment (hpatro) [inline src/t_stream.c]:
```suggestion
                    } else if (mode == 2) { /* ACKED */
```

File: src/t_stream.c [hunk 9]
```diff
@@ -3558,6 +3558,262 @@ void xdelCommand(client *c) {
...
+                        /* Target group hasn't received it: can't ack.
+                         * Non-target hasn't claimed it: acked but can't delete yet. */
+                        resps[j] = is_target ? -1 : 2;
+                    }
+                    continue;
+                }
+
+                void *result;
+                if (raxFind(cg->pel, buf, sizeof(buf), &result)) {
+                    if (is_target || mode == 1 /* DELREF */) {
```
Comment (hpatro) [inline src/t_stream.c]:
```suggestion
                    if (is_target || mode == 1 ) { /* DELREF */
```

File: src/t_stream.c [hunk 10]
```diff
@@ -3558,6 +3558,262 @@ void xdelCommand(client *c) {
...
+
+                streamID *id = &ids[j];
+                unsigned char buf[sizeof(streamID)];
+                streamEncodeID(buf, id);
+
+                /* Message hasn't been delivered to this group yet. */
+                if (streamCompareID(id, &cg->last_id) > 0) {
+                    if (!streamEntryExists(s, id)) {
+                        resps[j] = -1;
+                    } else if (mode == 2 /* ACKED */) {
```
Comment (hpatro) [inline src/t_stream.c]:
```suggestion
                    } else if (mode == 2) { /* ACKED */
```

File: src/t_stream.c [hunk 11]
```diff
@@ -3558,6 +3558,262 @@ void xdelCommand(client *c) {
...
+                }
+
+                streamID *id = &ids[j];
+                unsigned char buf[sizeof(streamID)];
+                streamEncodeID(buf, id);
+
+                /* Message hasn't been delivered to this group yet. */
+                if (streamCompareID(id, &cg->last_id) > 0) {
+                    if (!streamEntryExists(s, id)) {
+                        resps[j] = -1;
```
Comment (hpatro) [inline src/t_stream.c]: Why do we check if the entry exists after the streamCompareID check? Shouldn't we flip it?

File: src/t_stream.c [hunk 12]
```diff
@@ -3558,6 +3558,262 @@ void xdelCommand(client *c) {
...
+sync:
+    /* Update the stream's first ID. */
+    if (deleted) {
+        if (s->length == 0) {
+            s->first_id.ms = 0;
+            s->first_id.seq = 0;
+        } else if (first_entry) {
+            streamGetEdgeID(s, 1, 1, &s->first_id);
+        }
+    }
```
Comment (hpatro) [inline src/t_stream.c:3814]: Lot of the code here is duplicated across xdel. Will be good if we can refactor them into one 
 method.

File: src/t_stream.c [hunk 13]
```diff
@@ -3558,6 +3558,262 @@ void xdelCommand(client *c) {
...
+                        raxRemove(cg->pel, buf, sizeof(buf), NULL);
+                        raxRemove(nack->consumer->pel, buf, sizeof(buf), NULL);
+                        streamFreeNACK(nack);
+                        acked++;
+                    } else if (mode == 2 /* ACKED */) {
+                        resps[j] = 2; /* Another group still has a pending reference. */
+                    }
+                    /* KEEPREF (!is_target && mode == 0): leave other groups' PEL alone. */
+                } else {
+                    if (is_target && mode == 2 /* ACKED */) {
```
Comment (hpatro) [inline src/t_stream.c]:
```suggestion
                    if (is_target && mode == 2) { /* ACKED */
```

**Review by hpatro — COMMENTED** [review]
Thanks for the patience @nickiaq. Got to this finally. 

The code mostly looks good. Have some nitpicks. There is a quite few branching across the code block due to the nature of the command. Need to look at the tests next. Would love additional eyes on this. @soloestoy would you be interested in taking a pass at this?

File: src/t_stream.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (nickiaq) [reply src/t_stream.c]: Yes! This was a big issue. I’ve added a fix for this in a2164da5b by checking the target group first and then only acting on other groups / deleting the message if the message was in the target group’s PEL in the first place.

File: src/t_stream.c (hunk 4 above)
Comment (nickiaq) [reply src/t_stream.c:3593]: The mode is optional and defaults to `KEEPREF` if omitted. In this case, we’d get `IDS` in this argument position instead of the mode. This `else if` is handling the default mode case vs an invalid mode.

For example, the test `tests/unit/type/stream.tcl#L689` [here](https://github.com/valkey-io/valkey/pull/3466/changes#diff-9134b3f9730b7eccb57ce324638a9fcdc29ef498f0f0358c0a5394f1783dd385R689)) shows this behavior. Versus: there’s a test for invalid mode in `tests/unit/type/stream.tcl#L958` [here](https://github.com/valkey-io/valkey/pull/3466/changes#diff-9134b3f9730b7eccb57ce324638a9fcdc29ef498f0f0358c0a5394f1783dd385R958).

I saw below that you mentioned cleaning up duplicate code. I was thinking of doing that here too, because this exact switch is duplicated in both `XDELEX` and here. But, since I put those as separate PR’s (in case there’s interest in merging one and waiting or not merging the other), I was thinking I’d wait until they’re both merged to create a shared utility function used by both.

File: src/t_stream.c (hunk 5 above)
Comment (nickiaq) [reply src/t_stream.c]: I've adjusted as you describe [here](https://github.com/valkey-io/valkey/pull/3466/changes#diff-9f89868d206f9dbc0e415ba1f4f00b88f3c8766fafe346aa0744ae64024e052bR3604) and adjusted tests to match the standardized error that first check returns.

```
    /* Parse and validate numids: must be a positive integer. */
    long long id_count;
    if (getLongLongFromObjectOrReply(c, c->argv[argi], &id_count, NULL) == C_ERR) {
        return;
    }
    if (id_count <= 0) {
        addReplyError(c, "The IDs argument must be a positive integer");
        return;
    }
    argi++; /* past numids */
```

File: src/t_stream.c (hunk 6 above)
Comment (nickiaq) [reply src/t_stream.c:3624]: If this is referring to the check at the top of the function on line 3571 ([here](https://github.com/valkey-io/valkey/pull/3466/changes#diff-9f89868d206f9dbc0e415ba1f4f00b88f3c8766fafe346aa0744ae64024e052bR3571)), I think that only checks that the object is a stream if the key has a value set to something.

```
    robj *o = lookupKeyRead(c->db, c->argv[1]);
    if (o) {
        if (checkType(c, o, OBJ_STREAM)) return; /* Type error. */
        group = streamLookupCG(objectGetVal(o), objectGetVal(c->argv[2]));
    }
```

If there is no value at that key (ie. `NULL`), it will still continue through to the check on line 3621 referenced in the review comment. Tracing `lookupKeyRead` through to `lookupKey` in `src/db.c` I see the comment: *Lookup a key for read or write operations, or* ***return NULL if the key is not found***. In this case where the key isn’t found, `o == NULL` so the check at the top doesn’t catch it.

The problem with pushing this error check further up is that we need the ID count to be able to return -1 for each ID if we want to be compatible with Redis here. So, we need to parse the command first even if the stream doesn’t exist. Whereas the check at the top exists to return `WRONGTYPE` error if the key is set to something.

(Or, happy to correct if you’re referring to another check that I’m missing or if there’s some functionality of `lookupKeyRead` that I’m not understanding right here)

File: src/t_stream.c (hunk 7 above)
Comment (nickiaq) [reply src/t_stream.c]: Applied this suggestion in 75983f740 and subsequent commits.

File: src/t_stream.c (hunk 8 above)
Comment (nickiaq) [reply src/t_stream.c]: Applied this suggestion in 75983f740 and subsequent commits.

File: src/t_stream.c (hunk 9 above)
Comment (nickiaq) [reply src/t_stream.c]: Applied this suggestion in 75983f740 and subsequent commits.

File: src/t_stream.c (hunk 10 above)
Comment (nickiaq) [reply src/t_stream.c]: Applied this suggestion in 75983f740 and subsequent commits. (Sorry there were a lot of these)

File: src/t_stream.c (hunk 13 above)
Comment (nickiaq) [reply src/t_stream.c]: Here too; Applied this suggestion in 75983f740 and subsequent commits.

File: src/t_stream.c (hunk 11 above)
Comment (nickiaq) [reply src/t_stream.c]: Initially, I set this up to have the `streamCompareID(id, &cg->last_id)` check first here because it’s a cheaper check than the other. From what I could tell reviewing the stream state tracking, `cg->last_id` is the ID of the latest stream message delivered to that consumer group. So I was thinking that means it’s not possible that the stream message is in the PEL of the group if its ID is larger than the latest thing delivered. Whereas `streamEntryExists` seems to walk the radix tree of stream messages to check if it exists.

But in researching your comment here, I realized the one exception to this is if you run `XGROUP SETID` to manually move the `last_id` to before a message in the PEL. In that case, the cheaper check and `continue` would produce incorrect result because the message in PEL is actually after the `last_id`.

So I’ve included a fix as you’ve described to be more defensive in commit: 75983f740

File: src/t_stream.c (hunk 12 above)
Comment (nickiaq) [reply src/t_stream.c:3814]: Should I do that cleanup in this PR? Or, should I send another PR after this is merged or based on this branch?

**Comment by nickiaq** [comment]
Thanks for the review! There was lots here to tackle and it was fun digging into the source to figure out some of the more nuanced ones here.

I’ve added a note on each review comment above. I had one question ([here](https://github.com/valkey-io/valkey/pull/3466#discussion_r3642792008)) and 2 spots where I provided a bit more context ([here](https://github.com/valkey-io/valkey/pull/3466#discussion_r3642776673) and [here](https://github.com/valkey-io/valkey/pull/3466#discussion_r3642786845)) on what it's trying to do.

In addition to the fixes above, I’ve also implemented something to handle this comment (https://github.com/valkey-io/valkey/pull/3467#discussion_r3629794172) from the XDELEX PR regarding deferring response enqueueing until after message deletion, `server.dirty`, and signaling. See ea8a26a6f.

File: src/commands/xackdel.json [hunk 14]
```diff
@@ -0,0 +1,97 @@
+{
+    "XACKDEL": {
+        "summary": "Acknowledge and (if possible) delete stream message(s).",
+        "complexity": "O(1)",
+        "group": "stream",
+        "since": "10.0.0",
```
Comment (sushilpaneru1) [inline src/commands/xackdel.json]: 9.2 ?

File: src/t_stream.c (hunk 4 above)
Comment (zuiderkwast) [reply src/t_stream.c]: Ack. But we have the check for IDS just below which returns syntax error:

```C
    /* Expect IDS token. */
    if (strcasecmp(objectGetVal(c->argv[argi]), "IDS") != 0) {
        addReplyError(c, "syntax error");
        return;
    }
```

Therefore, this check seems redundant and we can just skip the else clause.


```suggestion
    }
```

File: src/t_stream.c [hunk 15]
```diff
@@ -3558,6 +3558,285 @@ void xdelCommand(client *c) {
...
+        argi += 1;
+        mode = 2;
+    } else if (strcasecmp(objectGetVal(c->argv[argi]), "IDS") != 0) {
+        addReplyError(c, "xackdel mode not recognized");
+        return;
+    }
+
+    /* Expect IDS token. */
+    if (strcasecmp(objectGetVal(c->argv[argi]), "IDS") != 0) {
+        addReplyError(c, "syntax error");
```
Comment (zuiderkwast) [inline src/t_stream.c]:
```suggestion
        addReplyErrorObject(c, shared.syntaxerr);
```

File: src/t_stream.c [hunk 16]
```diff
@@ -3558,6 +3558,285 @@ void xdelCommand(client *c) {
...
+
+    /* Validate numids matches remaining arg count. */
+    long long actual_ids = c->argc - argi;
+    if (id_count > actual_ids) {
+        addReplyError(c, "numids parameter must match the number of IDs provided");
+        return;
+    } else if (id_count < actual_ids) {
+        addReplyError(c, "syntax error");
+        return;
+    }
```
Comment (zuiderkwast) [inline src/t_stream.c]: I don't think we need two different errors for these cases. Simply syntax error seems enough. Simpler code.

```suggestion
    if (id_count != actual_ids) {
        addReplyErrorObject(c, shared.syntaxerr);
        return;
    }
```

**Review by zuiderkwast — COMMENTED** [review]
I took a pass, trying to get this reviewed and merged before 9.2 RC1, next week. The logic looks correct. I have only minor comments. It should be close to merge.

If you can refactor to avoid the duplicated code with XDEL, it would be very good. We should not introduce technical dept and then plan to refactor it later, because that often means it never happens.

File: src/t_stream.c (hunk 6 above)
Comment (zuiderkwast) [reply src/t_stream.c:3618]: If o is NULL, groups is also NULL, because it's initialized to NULL and only change if o exists. Thus, it should be enough to check group. But it doesn't matter much to me if we check both or just one here.

File: src/t_stream.c (hunk 12 above)
Comment (zuiderkwast) [reply src/t_stream.c:3808]: Yes please. We shouldn't introduce duplicated code and push technical dept to the future.

**Comment by nickiaq** [comment]
Thanks @zuiderkwast for the review. Also, changed the command JSON to `"since": "9.2.0"` if that plan is to get this whipped into shape for merge before that RC.

In addition to the commit here, I've taken your larger advice to create a new PR (#4629) that combines this with #3467 `XDELEX` and adds shared helpers to reduce duplicated code across `XDEL`, `XDELEX` , and this.

**Comment by nickiaq** [comment]
If #4629 looks like the right direction for your comment about deduplicating code, then I can close this and #3467.

**Comment by hpatro** [comment]
Thanks @zuiderkwast for the help with the reviews. I'm in inline with the feedback to merge these two together and avoid lot's of redundant boilerplate code.

**Comment by nickiaq** [comment]
Closing this in favor of #4629!
